### PR TITLE
dissect: add new --with command for excuting commands with a DDI mounted

### PR DESCRIPTION
**Pull request by poettering**
A useful addition to the systemd-dissect tool.

Usecase:

```
systemd-dissect --with foo.raw tar cz . > foo.tar.gz
```

for converting a DDI to a tarball.

### CODE REVIEW

File: src/dissect/dissect.c [hunk 1]
```diff
@@ -126,12 +132,50 @@ static int help(void) {
         return 0;
 }
 
+static int patch_argv(int *argc, char ***argv, char ***buf) {
+        _cleanup_free_ char **l = NULL;
+        char **e;
+
```
Comment (bluca) [inline src/dissect/dissect.c:138]: sprinkle some assert fairy dust

File: src/dissect/dissect.c [hunk 2]
```diff
@@ -970,6 +1037,99 @@ static int action_umount(const char *path) {
...
+        created_dir = TAKE_PTR(temp);
+
+        r = dissected_image_mount_and_warn(m, created_dir, UID_INVALID, UID_INVALID, arg_flags);
+        if (r < 0)
+                return r;
+
+        mounted_dir = TAKE_PTR(created_dir);
+
+        if (chdir(mounted_dir) < 0)
```
Comment (bluca) [inline src/dissect/dissect.c]: any reason this is not done in the child process?

**Review by bluca — CHANGES_REQUESTED** [review]
some shellcheck warnings to fix

File: test/units/testsuite-50.sh [hunk 3]
```diff
@@ -41,6 +41,14 @@ systemd-dissect --json=short "${image}.raw" | grep -q -F '{"rw":"ro","designator
 systemd-dissect "${image}.raw" | grep -q -F "MARKER=1"
 systemd-dissect "${image}.raw" | grep -q -F -f <(sed 's/"//g' "$os_release")
 
+systemd-dissect --list "${image}.raw" | grep -q '^etc/os-release$'
+
+read SHA256SUM1 GARBAGE < <(systemd-dissect --copy-from "${image}.raw" etc/os-release | sha256sum)
+test "$SHA256SUM1" != ""
+read SHA256SUM2 GARBAGE < <(systemd-dissect --read-only --with "${image}.raw" sha256sum etc/os-release)
```
Comment (mrc0mmand) [inline test/units/testsuite-50.sh]:
```suggestion
read -r SHA256SUM2 _ < <(systemd-dissect --read-only --with "${image}.raw" sha256sum etc/os-release)
```

ditto

File: test/units/testsuite-50.sh [hunk 4]
```diff
@@ -41,6 +41,14 @@
 systemd-dissect "${image}.raw" | grep -q -F "MARKER=1"
 systemd-dissect "${image}.raw" | grep -q -F -f <(sed 's/"//g' "$os_release")
 
+systemd-dissect --list "${image}.raw" | grep -q '^etc/os-release$'
+
+read SHA256SUM1 GARBAGE < <(systemd-dissect --copy-from "${image}.raw" etc/os-release | sha256sum)
```
[the remark this answers is not in the document]
Comment (mrc0mmand) [reply test/units/testsuite-50.sh]:
```suggestion
read -r SHA256SUM1 _ < <(systemd-dissect --copy-from "${image}.raw" etc/os-release | sha256sum)
```

should make shellcheck happy

File: test/units/testsuite-50.sh (hunk 3 above)
Comment (mrc0mmand) [reply test/units/testsuite-50.sh]: Not sure how I managed to post this one outside of the shellcheck warning comment, oh well.

File: src/dissect/dissect.c (hunk 2 above)
Comment (poettering) [reply src/dissect/dissect.c]: no reason. maybe i am getting old, dunno!

**Comment by poettering** [comment]
Force pushed a new version, hopefully addressing all issues. PTAL.

**Review by bluca — APPROVED** [review]

**Pull request merged**